journal/src/main/java/io/zeebe/journal/file/SegmentedJournal.java

```diff
@@ -147,7 +160,7 @@ public long getFirstIndex() {
     if (!isEmpty()) {
       return segments.firstEntry().getValue().index();
     } else {
-      return 0;
+      return 1;
```

Comment:
Why? If it is empty, should we return an index for which no entry exists? Returning 0 sounds correct to me.

Reply:
If I recall correctly, there's some behaviour in Atomix which expects the first index to always be 1, and the last index to be 0 in that case. But it shouldn't be hardcoded and rather we should return the index of the first segment's descriptor (which happens to be 1).

Reply:
Then I would then first verify if the usage of #getFirstIndex is correct. For me it doesn't seem to be semantically correct to return a valid index if the journal is empty.

Reply:
There's some assumptions in the Raft code around that as well - see

RaftContext
```java
private void verifySnapshotLogConsistent() {
    final long currentSnapshotIndex = getCurrentSnapshotIndex();
    if ((currentSnapshotIndex <= 0L && raftLog.getFirstIndex() != 1)
        || (currentSnapshotIndex > 0L
            && currentSnapshot.getIndex() + 1 < raftLog.getFirstIndex())) {
      // There is no snapshot, but the log has been compacted!
      throw new IllegalStateException(
          String.format(
              "Expected to find a snapshot at index >= log's first index %d, but found snapshot %d",
              raftLog.getFirstIndex(), currentSnapshotIndex));
    }
  }
```

Or when we initialize the `LeaderAppender` initial index (which is either the last index, or the first index when the log is empty):

```java
leaderIndex =
        raft.getLog().isEmpty() ? raft.getLog().getFirstIndex() : raft.getLog().getLastIndex() + 1;
```

If first index is 0, then the server doesn't start because later when we want to commit the initial entries (also in LeaderAppender):

```java
  public CompletableFuture<Long> appendEntries(final long index) {
    raft.checkThread();

    if (index == 0) {
      return appendEntries();
    }

    if (index <= raft.getCommitIndex()) {
      return CompletableFuture.completedFuture(index);
    }

    // If there are no other stateful servers in the cluster, immediately commit the index OR
    // If there are no other active members in the cluster, update the commit index and complete the
    // commit.
    // The updated commit index will be sent to passive/reserve members on heartbeats.
    if (raft.getCluster().getActiveMemberStates().isEmpty()) {
      final long previousCommitIndex = raft.getCommitIndex();
      raft.setCommitIndex(index);
      completeCommits(previousCommitIndex, index);
      return CompletableFuture.completedFuture(index);
    }

    // Only send entry-specific AppendRequests to active members of the cluster.
    return appendFutures.computeIfAbsent(
        index,
        i -> {
          for (final RaftMemberContext member : raft.getCluster().getActiveMemberStates()) {
            appendEntries(member);
          }
          return new CompletableFuture<>();
        });
  }
```

See the line where it says `index == 0`? When committing the initial entries, we pass the leaderIndex, in which case it does this special snowflake thing of sending a heartbeat :smile:

Reply:
Ok. In the leaderAppender it is already checking if it is empty before reading getFirstIndex, so it should be ok.  For the other check `currentSnapshotIndex <= 0L && raftLog.getFirstIndex() != 1)`, it should also add a check if the journal is empty. We can do it another PR.
For this PR, instead of returning 1, it should return firstSegment.index().